File: Backend/serverPython/controller/photo_controller.py
import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
import base64
import binascii
from flask import request
from serverPython.database.db import get_db_conn
import json
from serverPython.s3.s3 import upload_file

def uploadPicture(id):
    data = request.json
    nombre_foto = data.get('nombre_foto')
    album = data.get('album')
    foto = data.get('foto')
    url_foto = ""

    print(id, nombre_foto, album, foto)

    # Primero obtener el nombre del usuario
    query = f"SELECT id_usuario, usuario FROM usuario WHERE id_usuario = {id};"
    print(query)

    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query)
        result = cursor.fetchone()
        
        cursor.close()
        db_conn.close()

        nombre_usuario = result[1]
    except Exception as error: 
        print(error)
        return {"status": 400, "mensaje": "Datos no encontrados"}
    
    # Construir la URL de la foto
    url_foto = f"https://practica2-semi1-b-2s2024-imageness-g4.s3.amazonaws.com/Fotos_Publicadas/{nombre_usuario}{nombre_foto}.jpg"

    # Eliminar el prefijo si existe
    if foto.startswith("data:image/jpeg;base64,"):
        foto = foto.replace("data:image/jpeg;base64,", "")

    # Subir al bucket S3
    try:
        image_data = base64.b64decode(foto)  # Asegúrate de que la cadena esté en un formato correcto
        upload_file(image_data, nombre_foto, nombre_usuario)
    except (binascii.Error, ValueError) as e:
        print(f"Error al decodificar base64: {e}")
        return {"status": 400, "mensaje": "Error en la cadena base64"}

    query = f"CALL sp_foto({id}, '{album}', NULL, 'C', '{nombre_foto}', '{url_foto}', @codigo, @mensaje);"
    print(query)

    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query)
        cursor.execute("SELECT @codigo, @mensaje;")
        result = cursor.fetchone()
        codigo = result[0]
        mensaje = result[1]

        cursor.close()
        db_conn.close()
        return {"codigo": codigo, "mensaje": mensaje}
    except Exception as error: 
        print(error)
        return {"status": 400, "mensaje": "Error al subir la foto"}

def getPhotos(id):
    data = request.json
    id_album = data.get('id_album')

    query = f"SELECT id_foto, nombre_foto, url_foto FROM foto WHERE id_usuario = {id} AND id_album = {id_album};"

    db_conn = get_db_conn()
    try:
        cursor = db_conn.cursor()
        cursor.execute(query)
        result_sets = cursor.fetchall()
        print("result sets: ", result_sets)
        
        cursor.close()
        db_conn.close()

        data = [{"id_foto": row[0], "nombre_foto": row[1], "url_foto": row[2]} for row in result_sets]
        json_data = json.dumps(data, indent=4)

        return json_data
    except Exception as error: 
        print(error)
        return {"status": 400, "mensaje": "No se pudo obtener la información."}

File: Backend/server/controller/photo.controller.js
const query = require("../database/query");
const s3 = require("../s3/s3");

// Subir Foto
// CALL sp_foto(id_usuario, id_album, id_foto=NULL, 'C', 'nombre_foto', 'url', @codigo, @mensaje);
const uploadPicture = async (req, res) => {
  var user_id = req.params["id"];
  const { nombre_foto, album, foto } = req.body;
  let nombre_usuario;

  //obtener nombre de usuario
  const datos = await query.execute_(
    "call sp_usuario(@mensaje, ?, NULL, NULL, NULL, NULL, 'R');",
    [user_id]
  );
  if (datos.err) {
    return res.status(400).json(datos.err);
  } else {
    nombre_usuario = datos[0][0][0]["p_usuario"];
    console.log(nombre_usuario);
  }

  let url_foto =
    "https://practica2-g4-imagenes.s3.us-east-2.amazonaws.com/" +
    "fotos/" +
    nombre_usuario +
    nombre_foto +
    ".jpg";

  // BUCKET S3
  s3.uploadPicture(foto, nombre_foto, nombre_usuario);

  const [outcome] = await query.execute_(
    "CALL sp_foto(?, ?, NULL, 'C', ?, ?, @codigo, @mensaje); select @codigo, @mensaje;",
    [user_id, album, nombre_foto, url_foto]
  );
  console.log(url_foto);
  if (outcome.err) {
    return res.status(400).json(outcome.err);
  } else {
    return res
      .status(200)
      .json({
        codigo: outcome[1][0]["@codigo"],
        mensaje: outcome[1][0]["@mensaje"],
      });
  }
};

// Ver fotos de un album
// CALL practica1.sp_foto(id_usuario, id_album, id_foto, 'R', nombre_foto , url_foto , @codigo, @mensaje);
// CALL practica1.sp_foto(1, 1, NULL, 'R', NULL , NULL , ?, ?);
const getPhotos = async (req, res) => {
  var user_id = req.params["id"];
  const { id_album } = req.body;
  const [outcome] = await query.execute_(
    "CALL sp_foto(?, ?, NULL, 'R', NULL , NULL , @codigo, @mensaje);",
    [user_id, id_album]
  );
  if (outcome.err) {
    return res.status(400).json(outcome.err);
  } else {
    return res.status(200).json(outcome[0]);
  }
};

module.exports = {
  uploadPicture,
  getPhotos,
};
